### PR TITLE
feat: Batch with flusher

### PR DESCRIPTION
**Pull request by catShaark**
Closes: #652 

This PR introduces `BatchWithFlusher` which implement `cosmos-db.Batch` interface so that it could be used within `nodedb`

```
type BatchWithFlusher struct {
	db                      dbm.DB    // This is only used to create new batch
	batch                   dbm.Batch // Batched writing buffer.
	batchSizeFlushThreshold int       // The maximum size of the batch in bytes before it gets flushed to disk
}
```

### CODE REVIEW

**Comment by tac0turtle** [comment]
any chance we can get a benchmark with goleveldb added here.

File: batch_test.go
Comment (tac0turtle) [inline batch_test.go:26]: testing.B should have a tempDir now in go 1.18, can we try using that? it would remove the above cleanupDir as well

File: batch.go
Comment (p0mvn) [inline batch.go:7]:
```suggestion
// BatchWithFlusher is a wrapper
```

File: batch.go
Comment (p0mvn) [inline batch.go:23]: Can we make this a configurable parameter with the default value set to this? We should also link the source of this information.

Let me know if I can help with finding the source

File: batch.go
Comment (p0mvn) [inline batch.go:20]: Is this DB backend specific?

If that's the case, I think we should utilize this depending on the DB backend itself.

File: batch.go
Comment (p0mvn) [inline batch.go:50]: What do you think about changing this to the following flow:
1. b.batch.Set(key, value)
2. b.batch.GetByteSize()
3. if batch size >= b.batchSizeFlushThreshold, flush it to disk

This way, we can be agnostic of the `possibleEntryOverHead` that is db backend specific.

I think that exceeding `b.batchSizeFlushThreshold` is acceptable. We can just redefine this method to flush whenever the threshold is actually exceeded

File: batch.go
Comment (catShaark) [reply batch.go:23]: https://github.com/ethereum/go-ethereum/pull/15115 I think the refrence is here

File: batch.go
Comment (catShaark) [reply batch.go:50]: I just don't want to re-allocate the content of the batch when we exceed `b.batchSizeFlushThreshold`, since we've already pre-allocated the batch with `NewBatchWithSize`

File: batch_test.go
Comment (catShaark) [reply batch_test.go:35]: yesss

File: batch.go
Comment (p0mvn) [reply batch.go:23]: Amazing, thanks!

File: batch.go
Comment (p0mvn) [reply batch.go:20]: I'm concerned with exposing DB-backend-specific constraints on all other backends. Ideally, this complexity should be abstracted away under the backend implementation.

Imagine only rocksdb has this `possibleEntryOverhead` constraint and goleveldb doesn't.

Then, for goleveldb we would flush earlier than the threshold is actually reached due to over accounting for the `possibleEntryOverhead` that is rocks-backend specific

File: batch.go
Comment (p0mvn) [reply batch.go:50]: Why do you think that this would require redundant re-allocation? 

It seems to me that we only need to re-allocate after we flush the batch exceeding the threshold which is the behavior in the current design as well

Comment (catShaark) [reply batch.go:50]: it's because we specify a limit for the batch, we pre-allocate the batch with that limit size, and so when we unknowingly exceed that limit, (for example we set a key/value that make the batch size = limit + 10 byte ), the batch has to be allocated again since it is out of capacity

File: batch.go
Comment (p0mvn) [reply batch.go:20]: I see. Discussed offline more - so this only applies to a few backends, however, we overapply this to backends that do not have this overhead only when accounting for the last item in the batch.

As a result of the discussion, I agree to keep the implementation as is. 

Let's please add more detailed documentation explaining all the context, including:
- the backends to which this overhead applies, including links to the relevant parts in the codebase
- the overhead applies to every entry in the batch. However, we over-account for this overhead for backends that do not have this overhead only for the last entry in the batch.

File: batch.go
Comment (p0mvn) [reply batch.go:54]: Makes sense. Thank you

File: batch.go
Comment (p0mvn) [inline batch.go:74]: Do we need to ` b.batch.Close()` here, similar to `Set()`?

File: batch.go
Comment (p0mvn) [inline batch.go:53]: Can we encapsulate this logic into a helper function with tests and reuse the helper between `Set()` and `Delete()`?

**Comment by p0mvn** [comment]
LGTM once the remaining minor comments are addressed. I would like to see benchmark results before merging this.

It would also be great to see some profiling graphs before and after this change while a commit is in-progress.

**Comment by catShaark** [comment]
the benchmark shows that batch of larger size ( size 10MB) performs better

**Comment by tac0turtle** [comment]
> the benchmark shows that batch of larger size ( size 10MB) performs better

could you post the benchmarks? do all sizes perform better?

**Comment by catShaark** [comment]
```BenchmarkBatchWithFlusher/threshold=100000/backend=goleveldb-8         	1000000000	         0.2018 ns/op	       0 B/op	       0 allocs/op
BenchmarkBatchWithFlusher/threshold=1000000/backend=goleveldb-8        	1000000000	         0.2410 ns/op	       0 B/op	       0 allocs/op
BenchmarkBatchWithFlusher/threshold=10000000/backend=goleveldb-8       	1000000000	         0.1659 ns/op	       0 B/op	       0 allocs/op```

**Comment by tac0turtle** [comment]
is there a comparison to what lives on main already? @catShaark

**Comment by catShaark** [comment]
No but I can do that

**Comment by tac0turtle** [comment]
awesome, lets get that and this branch updated to master then we can merge

**Comment by ValarDragon** [comment]
I haven't reviewed this PR, but I did the idea of this PR just for genesis, for fast nodes, and got a significant speedup.

Before:
image
After:
image

And GC time down by ~200 seconds as well.

**Review by tac0turtle — APPROVED** [review]

**Comment by tac0turtle** [comment]
@mergifyio backport release/v1.x.x

**Pull request merged**
